proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,190 @@
+
+1. default value?
+
+   Should we add nuget.org as a default value, so that customers who remove nuget.org as a package source are more easily able to use NuGetAudit?
+   Would using nuget.org as a default audit source cause undue concern from customers who don't want to use nuget.org as a package source, and the new tooling version starts making network requests to nuget.org?
+
+   If we want to make nuget.org a default/automatic audit source, how?
+   NuGet has tried before to use a tracking file to automatically add nuget's v3 URL as a package source to user-profile nuget.config files that don't have any package sources.
+   However, this caused a security issue when customers wanted to remove nuget.org as a default package source, and this code incorrectly re-added it once (the customer could successfully remove nuget.org a second time).
+   When we removed this auto-update to mitigate the security concern, it exposed a bug where multiple other apps created user-profile nuget.config files without any package sources, and many customers assumed it was a bug with NuGet.
+
+   Should we hard-code nuget.org's default URL in NuGet.Configuration, and require customers to use `<clear/>` if they don't want any audit source?
+
+   Should we just add it as a default in new user-profile nuget.config files, and just accept that existing developers won't get it by default?
+
+2. Stop using package source as audit source?
+
+   On one hand, using package sources as a defacto audit source makes it easier for customers to use the feature when they knowingly or unknowingly use a package source with a VDB.
+   On the other hand, overloading `<packageSources>` with multiple meanings is confusing and customers probably won't understand why sometimes they get vulnerable package information and sometimes they don't.
+   If we add nuget.org as a default audit source, and don't use packageSources when auditSources has a value, then using packageSources to get a VDB when auditSources is empty will be particularly confusing/unintuitive to customers.
+
+   Another challenge is that if we want to change the behavior it' will be a breaking change, which is typically discouraged.
+   However, given that nuget.org is the only know source with a VDB, if we make nuget.org a default audit source, then it will effectively be a non-breaking change.
+
+3. Tooling
+
+   How important is it to add CLI commands to manage audit sources?
+   How important is it to add a GUI in Visual Studio to manage audit sources?
+
+4. Restore summary
+
+   When you restore at normal verbosity, NuGet outputs a list of package sources that were used for restore. Should we add audit sources as well?
+   If so, presumably we should only output it if at least 1 project in the restore has NuGetAudit enabled, as outputting audit sources when NuGetAudit is disabled for all projects might confuse customers.
```

Comment:
Isn't this just a flat list that isn't correlated to projects? How will it indicate which projects were analyzed and which weren't?

Reply:
I think there's a misunderstanding. Since this is one of the "additional questions" that I'm going to delete before the spec is merged, I didn't bother putting effort into an example. But what I'm talking about is this:

```text
         NuGet Config files used:
             C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config
             C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config
             C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

         Feeds used:
             https://api.nuget.org/v3/index.json
             C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
             C:\Program Files\dotnet\library-packs
````

This question is about adding an additional bit:

```text
         Audit sources used:
             https://api.nuget.org/v3/index.json
```


> How will it indicate which projects were analyzed and which weren't?

This spec is intended to be specific about being able to audit sources which are not package sources. Other generic NuGetAudit features were intentionally out of scope.

Your suggestion overlaps greatly with something already proposed in the original NuGetAudit (whole of feature) spec: https://github.com/NuGet/Home/blob/dev/proposed/2022/vulnerabilities-in-restore.md#command-line-restore-output

We also have a (private) tracking issue: https://github.com/NuGet/Client.Engineering/issues/2340  Why is it a private tracking issue? I don't know. Seems like an oversight now. Customers should be able to upvote the idea to give us feedback they think it's more or less important than other features.

Reply:
> if at least 1 project in the restore has NuGetAudit enabled,

When we output those "Feeds used", those are applying to a solution.
Similarly, the "NuGet Config files used:" are solution-level concepts.
Can you explain how a customer will know that "Audit sources used:" is relevant to a project as your comment suggests?

Reply:
Since audit sources can't be defined by MSBuild, then whatever set of audit sources are listed in the nuget.config apply to all projects. It's just that projects that disable the entire NuGetAudit feature won't use them, but they're still the "defined" audit sources. This is not conceptually different to package sources ("feeds used") when a project doesn't use any packages (either directly or transitively).

I believe your comment/question is more about "how do customers get feedback about which projects have NuGetAudit enabled or disabled". This feature spec (aimed at using nuget.org vulnerability database without also using it as a package source) is orthogonal to that question. I can understand why you're bringing it up, since this is discussing showing something in the restore summary that wasn't being shown earlier. But I feel like this feature doesn't make the "do individual projects enable/disable audit" any better or worse.

The original NuGetAudit spec from last year included generalized restore summary (including a proposed output format), although it was moved to the "future possibilies" section. There's also [this tracking issue](https://github.com/NuGet/Client.Engineering/issues/2340) that I linked to in a few other "conversations" in this PR.

Reply:
> I believe your comment/question is more about "how do customers get feedback about which projects have NuGetAudit enabled or disabled"

Yes, exactly. I chose "indicate" and you chose "get feedback", but we're thinking the same.
# 2340 seems to address my concern.